### ISSUE TITLE
[desktop_multi_window] Can it work with window_manager?

### ISSUE THREAD

**Issue by Merrit** [MixinNetwork/flutter-plugins#137]
At the moment if I try to access `window_manager` from a sub-window it throws an exception:

> MissingPluginException (MissingPluginException(No implementation found for method ensureInitialized on channel window_manager))

**Comment by Narek6556** [comment]
Read this one 
https://github.com/MixinNetwork/flutter-plugins/issues/58
You need to add some changes in MainFlutterWindow.swift file for mac os and flutter_window.cpp for windows

**Comment by krida2000** [comment]
@Narek6556 any ideas how it can be done automatically? In [multi_window](https://pub.dev/packages/multi_window) no problems with it.

**Comment by Narek6556** [comment]
@krida2000 I don’t know how we can do it automatically.

**Comment by Merrit** [comment]
Thanks! I got it mostly working, only thing that seems to be an issue is that I can't set a transparent background in a sub-window; works fine in the main window. Is this also a known issue?

[Minimal reproducable example](https://github.com/Merrit/flutter_transparent_temp)

**Comment by angelocordero** [comment]
> Thanks! I got it mostly working, only thing that seems to be an issue is that I can't set a transparent background in a sub-window; works fine in the main window. Is this also a known issue?
> 
> [Minimal reproducable example](https://github.com/Merrit/flutter_transparent_temp)

How did you make it work? Link to your example is no longer working

**Comment by Merrit** [comment]
> How did you make it work? Link to your example is no longer working

The solution was the quoted link [here](https://github.com/MixinNetwork/flutter-plugins/issues/58#issuecomment-1060371162).

Example of my usage on Linux:

`<project_root>/linux/my_application.cc`

- add imports

```c
#include "desktop_multi_window/desktop_multi_window_plugin.h"
#include <screen_retriever/screen_retriever_plugin.h>
#include <tray_manager/tray_manager_plugin.h>
#include <window_manager/window_manager_plugin.h>
#include <window_size/window_size_plugin.h>
```

- at the end of `my_application_activate()` populate the callback

```c
// ...

fl_register_plugins(FL_PLUGIN_REGISTRY(view));

  desktop_multi_window_plugin_set_window_created_callback([](FlPluginRegistry* registry){
    g_autoptr(FlPluginRegistrar) window_manager_registrar =
        fl_plugin_registry_get_registrar_for_plugin(registry, "WindowManagerPlugin");
    window_manager_plugin_register_with_registrar(window_manager_registrar);
    g_autoptr(FlPluginRegistrar) window_size_registrar =
      fl_plugin_registry_get_registrar_for_plugin(registry, "WindowSizePlugin");
    window_size_plugin_register_with_registrar(window_size_registrar);
  });

  gtk_widget_grab_focus(GTK_WIDGET(view));
```

**Comment by rashcroft7** [comment]
I managed to get it working on Windows with the following in flutter_window.cpp:

- Imports
`#include "desktop_multi_window/desktop_multi_window_plugin.h"
#include <window_manager/window_manager_plugin.h>`

- Code

`  RegisterPlugins(flutter_controller_->engine());
   DesktopMultiWindowSetWindowCreatedCallback([](void *controller) {
     auto *flutter_view_controller =
         reinterpret_cast<flutter::FlutterViewController *>(controller);
     auto *registry = flutter_view_controller->engine();
     WindowManagerPluginRegisterWithRegistrar(
         registry->GetRegistrarForPlugin("WindowManagerPlugin"));
   });
  SetChildContent(flutter_controller_->view()->GetNativeWindow());`

I am now having an issue where the application crashes when the child window is closed.  This only applies when using Window Manager.

**Comment by whaleHan** [comment]
> 我设法让它在 Windows 上运行，flutter_window.cpp年具有以下功能：
> 
> * 进口
> 
> ```c++
> #include "desktop_multi_window/desktop_multi_window_plugin.h"
> #include <window_manager/window_manager_plugin.h>
> ```
> 
> * 法典
> 
> ```c++
> RegisterPlugins(flutter_controller_->engine());
> DesktopMultiWindowSetWindowCreatedCallback([](void *controller) {
>   auto *flutter_view_controller =
>       reinterpret_cast<flutter::FlutterViewController *>(controller);
>   auto *registry = flutter_view_controller->engine();
>   WindowManagerPluginRegisterWithRegistrar(
>       registry->GetRegistrarForPlugin("WindowManagerPlugin"));
> });
> SetChildContent(flutter_controller_->view()->GetNativeWindow());
> ```
> 
> 我现在遇到一个问题，当子窗口关闭时应用程序崩溃。这仅适用于使用窗口管理器时。
I'm also having this problem with WindowManager and MultiWindow, is there a solution?

**Comment by annd22** [comment]
is there any update for this?

**Comment by abldabl** [comment]
> is there any update for this?
did you find solution?

**Comment by 769755893** [comment]
the reason of crashing when closing sub window is the global engine channel is overriden by the new engine, so you just move the global engine channel to the manager class and hold it in inner class on each window instance.

**Comment by abldabl** [comment]
can you give some code for this, please?
> the reason of crashing when closing sub window is the global engine channel is overriden by the new engine, so you just move the global engine channel to the manager class and hold it in inner class on each window instance.

**Comment by 769755893** [comment]
`void WindowManagerPlugin::_EmitEvent(std::string eventName) {
  flutter::EncodableMap args = flutter::EncodableMap();
  args[flutter::EncodableValue("eventName")] =
      flutter::EncodableValue(eventName);
  if (channel != nullptr) {
      std::cout << "channel not nullptr start invoke event on new engine" << std::endl;
      try {
          channel->InvokeMethod("onEvent",
              std::make_unique<flutter::EncodableValue>(args));
      }
      catch (std::exception& e) {
          std::cout << e.what() << std::endl;
      }
  }
}`
you can see it the main reason crashing is the pre sub window 's method channel is nullptr, just hold it and move the channel variables to each window instance classes.

**Comment by annd22** [comment]
> `void WindowManagerPlugin::_EmitEvent(std::string eventName) { flutter::EncodableMap args = flutter::EncodableMap(); args[flutter::EncodableValue("eventName")] = flutter::EncodableValue(eventName); if (channel != nullptr) { std::cout << "channel not nullptr start invoke event on new engine" << std::endl; try { channel->InvokeMethod("onEvent", std::make_unique<flutter::EncodableValue>(args)); } catch (std::exception& e) { std::cout << e.what() << std::endl; } } }` you can see it the main reason crashing is the pre sub window 's method channel is nullptr, just hold it and move the channel variables to each window instance classes.

Just a little curious, how and where can u view the `cout` of the cpp native code in flutter, every I want to print something to debug in native code, I have to write it out to a local file and read it. Thanks!

**Comment by 769755893** [comment]
> 

use android studio to install flutter plugin and run project can do it, maybe vscode can do it as well.

**Comment by annd22** [comment]
Thank you, Im using vs code, maybe it only works on android studio. Let me try.

**Comment by abldabl** [comment]
i added analog code on linux. 
but i lost ability to call  DesktopMultiWindow.invokeMethod between window.

do you have similar issue?
i'm really stuck on multi window feature, and calling method between sub-window and main-window

> `void WindowManagerPlugin::_EmitEvent(std::string eventName) { flutter::EncodableMap args = flutter::EncodableMap(); args[flutter::EncodableValue("eventName")] = flutter::EncodableValue(eventName); if (channel != nullptr) { std::cout << "channel not nullptr start invoke event on new engine" << std::endl; try { channel->InvokeMethod("onEvent", std::make_unique<flutter::EncodableValue>(args)); } catch (std::exception& e) { std::cout << e.what() << std::endl; } } }`
> you can see it the main reason crashing is the pre sub window 's method channel is nullptr, just hold it and move the channel variables to each window instance classes.

**Comment by LucaCoduriV** [comment]
> > How did you make it work? Link to your example is no longer working
> 
> The solution was the quoted link [here](https://github.com/MixinNetwork/flutter-plugins/issues/58#issuecomment-1060371162).
> 
> Example of my usage on Linux:
> 
> `<project_root>/linux/my_application.cc`
> 
> * add imports
> 
> ```c
> #include "desktop_multi_window/desktop_multi_window_plugin.h"
> #include <screen_retriever/screen_retriever_plugin.h>
> #include <tray_manager/tray_manager_plugin.h>
> #include <window_manager/window_manager_plugin.h>
> #include <window_size/window_size_plugin.h>
> ```
> 
> * at the end of `my_application_activate()` populate the callback
> 
> ```c
> // ...
> 
> fl_register_plugins(FL_PLUGIN_REGISTRY(view));
> 
>   desktop_multi_window_plugin_set_window_created_callback([](FlPluginRegistry* registry){
>     g_autoptr(FlPluginRegistrar) window_manager_registrar =
>         fl_plugin_registry_get_registrar_for_plugin(registry, "WindowManagerPlugin");
>     window_manager_plugin_register_with_registrar(window_manager_registrar);
>     g_autoptr(FlPluginRegistrar) window_size_registrar =
>       fl_plugin_registry_get_registrar_for_plugin(registry, "WindowSizePlugin");
>     window_size_plugin_register_with_registrar(window_size_registrar);
>   });
> 
>   gtk_widget_grab_focus(GTK_WIDGET(view));
> ```

Hello, even wit this changes I still have this error:
`Unhandled Exception: MissingPluginException(No implementation found for method ensureInitialized on channel window_manager)`

Did you do something else ?

**Comment by Merrit** [comment]
I can get rid of the `MissingPluginException` error by following those instructions, but calling window_manager from a subwindow doesn't seem to do anything for me anymore.

**Comment by isaacy13** [comment]
> I can get rid of the `MissingPluginException` error by following those instructions, but calling window_manager from a subwindow doesn't seem to do anything for me anymore.

Like @Merrit, got rid of `MissingPluginException` by following instructions, but window_manager wasn't working in subwindows

Got it working after debug session on
- window_manager 0.3.2
- desktop_multi_window 0.2.0

**TLDR**:
issue was logic error with the order sub-window is created & sub-window attributes are set

make sure logic follows in order:
- use `DesktopMultiWindow.createWindow()` to create window
- use `WindowManager.instance` to set attributes (e.g.: title, fullscreen, etc.)
- use `WindowManager.instance.show()` to show window after attributes are set

⚠️ do NOT `WindowController.show()` using result from `DesktopMultiWindow.createWindow()` to show window -> causes  race condition

keep race conditions in mind when you're debugging/resolving this issue
- each sub-window creates a new thread
- logic must be in order, otherwise `window_manager` doesn't work on sub-windows

**Longer explanation:**
Two important blocks of code:

1. Somewhere, create sub-window
`DesktopMultiWindow.createWindow(...).then(window) { window.show() }`

2. When `DesktopMultiWindow.createWindow(...)` is called, `void main(args)` is run

```
void main(args) async {
   ...
   final isSubWindow = (args.firstOrNull == "multi_window")
   WidgetsFlutterBinding.ensureInitialized()
   await windowManager.ensureInitialized()
   ...
   if (!isSubWindow) {
      windowManager.waitUntilReadyToShow(windowOptions, () async {
          await windowManager.setTitle("Main Window")
          await windowManager.show()
      })
      runApp(MainWindowApp())
   }
   else {
      windowManager.waitUntilReadyToShow(windowOptions, () async {
          await windowManager.setTitle("Child Window")
      })
      runApp(SubWindowApp())
   }
}
```

The issue occurs when `window.show()` is called before attributes are set (e.g.: `await windowManager.setTitle("Child Window")`)

Making sure attributes are set before sub-window is shown solved issue for me

Correct usage:
1. Somewhere, create sub-window
`DesktopMultiWindow.createWindow(...)`

2. When `DesktopMultiWindow.createWindow(...)` is called, `void main(args)` is run
```
void main(args) async {
   ...
   final isSubWindow = (args.firstOrNull == "multi_window")
   WidgetsFlutterBinding.ensureInitialized()
   await windowManager.ensureInitialized()
   ...
   if (!isSubWindow) {
      runApp(MainWindowApp())
      windowManager.waitUntilReadyToShow(windowOptions, () async {
          await windowManager.setTitle("Main Window")
          await windowManager.show()
      })
   }
   else {
      runApp(SubWindowApp())
      windowManager.waitUntilReadyToShow(windowOptions, () async {
          await windowManager.setTitle("Child Window")
          await windowManager.show()
      })
   }
}
```

**Comment by isaacy13** [comment]
> I managed to get it working on Windows with the following in flutter_window.cpp:
> 
> * Imports
> 
> ```c++
> #include "desktop_multi_window/desktop_multi_window_plugin.h"
> #include <window_manager/window_manager_plugin.h>
> ```
> 
> * Code
> 
> ```c++
> RegisterPlugins(flutter_controller_->engine());
> DesktopMultiWindowSetWindowCreatedCallback([](void *controller) {
>   auto *flutter_view_controller =
>       reinterpret_cast<flutter::FlutterViewController *>(controller);
>   auto *registry = flutter_view_controller->engine();
>   WindowManagerPluginRegisterWithRegistrar(
>       registry->GetRegistrarForPlugin("WindowManagerPlugin"));
> });
> SetChildContent(flutter_controller_->view()->GetNativeWindow());
> ```
> 
> I am now having an issue where the application crashes when the child window is closed. This only applies when using Window Manager.

This solution worked perfectly for me, but if you always want sub-windows to have all plugins your main window has, this solution worked for me:

```
RegisterPlugins(flutter_controller_->engine());

// whenever DesktopMultiWindow creates a new window, do the following
DesktopMultiWindowSetWindowCreatedCallback([](void* controller) {
  // get view controller and engine of new window
  auto* flutter_view_controller =
  reinterpret_cast<flutter::FlutterViewController*>(controller);
  auto* registry = flutter_view_controller->engine();
  
  // register plugins with the new window
  RegisterPlugins(registry);
});

SetChildContent(flutter_controller_->view()->GetNativeWindow());
```

this way, you don't have to re-register for every plugin you will add in the future

**Comment by GangJust** [comment]
> > I managed to get it working on Windows with the following in flutter_window.cpp:
> > 
> > * Imports
> > 
> > ```c++
> > #include "desktop_multi_window/desktop_multi_window_plugin.h"
> > #include <window_manager/window_manager_plugin.h>
> > ```
> > 
> > 
> >     
> >       
> >     
> > 
> >       
> >     
> > 
> >     
> >   
> > 
> > * Code
> > 
> > ```c++
> > RegisterPlugins(flutter_controller_->engine());
> > DesktopMultiWindowSetWindowCreatedCallback([](void *controller) {
> >   auto *flutter_view_controller =
> >       reinterpret_cast<flutter::FlutterViewController *>(controller);
> >   auto *registry = flutter_view_controller->engine();
> >   WindowManagerPluginRegisterWithRegistrar(
> >       registry->GetRegistrarForPlugin("WindowManagerPlugin"));
> > });
> > SetChildContent(flutter_controller_->view()->GetNativeWindow());
> > ```
> > 
> > 
> >     
> >       
> >     
> > 
> >       
> >     
> > 
> >     
> >   
> > I am now having an issue where the application crashes when the child window is closed. This only applies when using Window Manager.
> 
> This solution worked perfectly for me, but if you always want sub-windows to have all plugins your main window has, this solution worked for me:
> 
> ```
> RegisterPlugins(flutter_controller_->engine());
> 
> // whenever DesktopMultiWindow creates a new window, do the following
> DesktopMultiWindowSetWindowCreatedCallback([](void* controller) {
>   // get view controller and engine of new window
>   auto* flutter_view_controller =
>   reinterpret_cast<flutter::FlutterViewController*>(controller);
>   auto* registry = flutter_view_controller->engine();
>   
>   // register plugins with the new window
>   RegisterPlugins(registry);
> });
> 
> SetChildContent(flutter_controller_->view()->GetNativeWindow());
> ```
> 
> this way, you don't have to re-register for every plugin you will add in the future

Thank you. Although the problem has been resolved, a new issue has emerged. When the sub window is closed (the main window is still running), it can cause the program to crash.

**Comment by isaacy13** [comment]
@GangJust 

This was also encountered & solved by `769755893` and `lycstar` earlier in the thread

Both of their solutions work great

- 769755893's solution
> `void WindowManagerPlugin::_EmitEvent(std::string eventName) { flutter::EncodableMap args = flutter::EncodableMap(); args[flutter::EncodableValue("eventName")] = flutter::EncodableValue(eventName); if (channel != nullptr) { std::cout << "channel not nullptr start invoke event on new engine" << std::endl; try { channel->InvokeMethod("onEvent", std::make_unique<flutter::EncodableValue>(args)); } catch (std::exception& e) { std::cout << e.what() << std::endl; } } }` you can see it the main reason crashing is the pre sub window 's method channel is nullptr, just hold it and move the channel variables to each window instance classes.


- lycstar's solution
> > `void WindowManagerPlugin::_EmitEvent(std::string eventName) { flutter::EncodableMap args = flutter::EncodableMap(); args[flutter::EncodableValue("eventName")] = flutter::EncodableValue(eventName); if (channel != nullptr) { std::cout << "channel not nullptr start invoke event on new engine" << std::endl; try { channel->InvokeMethod("onEvent", std::make_unique<flutter::EncodableValue>(args)); } catch (std::exception& e) { std::cout << e.what() << std::endl; } } }` you can see it the main reason crashing is the pre sub window 's method channel is nullptr, just hold it and move the channel variables to each window instance classes.
> 
> I don't understand what you mean, but if I try to do this, it won't crash, but it will cause only the main window's events to be triggered, and the child window's events won't be able to be triggered
> 
> ```c++
> void WindowManagerPlugin::_EmitEvent(std::string eventName) {
>   flutter::EncodableMap args = flutter::EncodableMap();
>   args[flutter::EncodableValue("eventName")] =
>       flutter::EncodableValue(eventName);
>   //add this to update channel
>   channel = std::make_unique<flutter::MethodChannel<flutter::EncodableValue>>(
>       registrar->messenger(), "window_manager",
>       &flutter::StandardMethodCodec::GetInstance());
>   //
>   channel->InvokeMethod("onEvent",
>                         std::make_unique<flutter::EncodableValue>(args));
> }
> ```


To summarize, it's a null dereference issue
- e.g.: `channel` variable is null & trying to call `InvokeMethod`
- When DesktopMultiWindow's sub-window is closed before main window, the MethodChannel `channel` is set to nullptr
- Symptom of WindowManager & DesktopMultiWindow not being 100% integrated together -- WindowManager closes MethodChannel when any window is closed (since there's normally only 1 window), but DesktopMultiWindow is still trying to make use of that same channel

How to solve it (for Windows)?
- Find `WindowManagerPlugin::_EmitEvent` in `window_manager_plugin.cpp` (though can differ, generally found in `C:/Users/username/AppData/Local/Pub/Cache/hosted/pub.dev/window_manager-x.x.x/windows/window_manager_plugin.cpp)`
- Use 769755893's solution (e.g.: if `channel` is nullptr, don't call `InvokeMethod`)
  - works, but events can no longer be emitted
- Or use lycstar's solution (e.g.: everytime an event is emitted, re-initialize the `channel` variable)
  - now events can be emitted, but we're reinitializing a `MethodChannel` each time we emit an event (not 100% perfect performance-wise)

Another note for both solutions is that whenever you upgrade the WindowManager package, you'll have to reimplement this solution

Overall, would recommend lycstar's approach so events can continue being passed
However, if you really wanted everything 100% perfect, you could consider forking the projects and putting in logic that integrates `WindowManager` with `DesktopMultiWindow`:
- whenever the **main window** (not any sub-windows) is closed, close the channel

Take my explanations with a grain of salt since it's been a week or two since I actually debugged

**Comment by eugeniz** [comment]
I showed several windows using desktop_multi_window package and want to hide/remove title bar of sub windows.
Is it possible to hide/remove them?
please help me.

**Comment by enfshin** [comment]
> I showed several windows using desktop_multi_window package and want to hide/remove title bar of sub windows. Is it possible to hide/remove them? please help me.

you_project\windows\flutter\ephemeral\.plugin_symlinks\desktop_multi_window\windows\flutter_window.cc

```
  HWND window_handle = CreateWindow(
      kFlutterWindowClassName, L"", WS_OVERLAPPEDWINDOW,
      Scale(target_point.x, scale_factor_), Scale(target_point.y, scale_factor_),
      Scale(1280, scale_factor_), Scale(720, scale_factor_),
      nullptr, nullptr, GetModuleHandle(nullptr), this);
```
change `WS_OVERLAPPEDWINDOW` to `WS_POPUP`

**Comment by enfshin** [comment]
> I showed several windows using desktop_multi_window package and want to hide/remove title bar of sub windows. Is it possible to hide/remove them? please help me.

if you follow above solutions, you can hide titlebar at subwindow. by setting 
`
WindowOptions windowOptions = const WindowOptions(
...
  titleBarStyle: TitleBarStyle.hidden,
);
`

**Comment by eugeniz** [comment]
Thanks for your reply.
It works!
And I am trying to sync state between main window and sub windows.
In desktop_multi_window, there is only option for invoking method from main window to sub windows.
No option for sub windows to main window. How to sync this?
I am using riverpod for state management and want to use same state in entire app.
If you have any solution, please let me know.

**Comment by isaacy13** [comment]
> Thanks for your reply. It works! And I am trying to sync state between main window and sub windows. In desktop_multi_window, there is only option for invoking method from main window to sub windows. No option for sub windows to main window. How to sync this? I am using riverpod for state management and want to use same state in entire app. If you have any solution, please let me know.

Talking from sub to main window should be the same as main to sub window

in MainWindow.dart, do:
```
DesktopMultiWindow.setMethodHandler((call, fromWindowId){
   switch(call.method) {
        "methodTriggeredBySubWindow":
        // write code here
        break
        default:
        // method not implemented!!!
        break
   }
})
```

in SubWindow.dart, do:
```
var mainWindowId = WindowController.main().windowId; // should always be 0
var result = DesktopMultiWindow.invokeMethod(
   mainWindowId,
   "methodTriggeredBySubWindow",
   {
      // args here
   }
)
```

Make sure the method handler is set before invoking it

----------------------------------------------------

Also, while doing this if you run into a MissingPluginException("no implementation found for channel"), then don't register DesktopMultiWindowPlugin on each sub-window
- devs only intended for it to be registered on main window, doing on each sub-window messes up events to main window

So unlike my [prior comment](https://github.com/MixinNetwork/flutter-plugins/issues/137#issuecomment-1704532577), would recommend registering each plugin separately in each sub-window

so in flutter_window.cpp (or its Linux/Mac counterparts):
```
// whenever DesktopMultiWindow creates a new window, do the following
DesktopMultiWindowSetWindowCreatedCallback([](void* controller) {
  // get view controller and engine of new window
  auto* flutter_view_controller =
    reinterpret_cast<flutter::FlutterViewController*>(controller);
  auto* registry = flutter_view_controller->engine();

  // register plugins with the new window
  // RegisterPlugins(registry); // don't do this
  WindowManagerPluginRegisterWithRegistrar(
        registry->GetRegistrarForPlugin("WindowManagerPlugin")); // do this with all your plugins, except for DesktopMultiWindow
  });
```

**Comment by nrbnlulu** [comment]
BTW you can find the current plugins that you use in  `generated_plugin_registrant.cc` for linux and windows

**Comment by zjw78123sc** [comment]
> > I can get rid of the `MissingPluginException` error by following those instructions, but calling window_manager from a subwindow doesn't seem to do anything for me anymore.
> 
> Like [@Merrit](https://github.com/Merrit), got rid of `MissingPluginException` by following instructions, but window_manager wasn't working in subwindows
> 
> Got it working after debug session on
> 
> * window_manager 0.3.2
> * desktop_multi_window 0.2.0
> 
> **TLDR**: issue was logic error where sub-window was shown before sub-window attributes were set
> 
> make sure logic follows in order:
> 
> * use `DesktopMultiWindow.createWindow()` to create window
> * use `WindowManager.instance` to set attributes (e.g.: title, fullscreen, etc.)
> * use `WindowManager.instance.show()` to show window after attributes are set
> 
> ⚠️ do NOT `WindowController.show()` using result from `DesktopMultiWindow.createWindow()` to show window -> causes race condition
> 
> keep race conditions in mind when you're debugging/resolving this issue
> 
> * each sub-window initializes asynchronously
> * logic must be in order, otherwise `window_manager` doesn't work on sub-windows
> 
> **Longer explanation:** Two important blocks of code:
> 
> 1. Somewhere, create sub-window
>    `DesktopMultiWindow.createWindow(...).then(window) { window.show() }`
> 2. When `DesktopMultiWindow.createWindow(...)` is called, `void main(args)` is run
> 
> ```
> void main(args) async {
>    ...
>    final isSubWindow = (args.firstOrNull == "multi_window")
>    WidgetsFlutterBinding.ensureInitialized()
>    await windowManager.ensureInitialized()
>    ...
>    if (!isSubWindow) {
>       windowManager.waitUntilReadyToShow(windowOptions, () async {
>           await windowManager.setTitle("Main Window")
>           await windowManager.show()
>       })
>       runApp(MainWindowApp())
>    }
>    else {
>       windowManager.waitUntilReadyToShow(windowOptions, () async {
>           await windowManager.setTitle("Child Window")
>       })
>       runApp(SubWindowApp())
>    }
> }
> ```
> 
> The issue occurs when `window.show()` is called before attributes are set (e.g.: `await windowManager.setTitle("Child Window")`)
> 
> Making sure attributes are set before sub-window is shown solved issue for me
> 
> Correct usage:
> 
> 1. Somewhere, create sub-window
>    `DesktopMultiWindow.createWindow(...)`
> 2. When `DesktopMultiWindow.createWindow(...)` is called, `void main(args)` is run
> 
> ```
> void main(args) async {
>    ...
>    final isSubWindow = (args.firstOrNull == "multi_window")
>    WidgetsFlutterBinding.ensureInitialized()
>    await windowManager.ensureInitialized()
>    ...
>    if (!isSubWindow) {
>       runApp(MainWindowApp())
>       windowManager.waitUntilReadyToShow(windowOptions, () async {
>           await windowManager.setTitle("Main Window")
>           await windowManager.show()
>       })
>    }
>    else {
>       runApp(SubWindowApp())
>       windowManager.waitUntilReadyToShow(windowOptions, () async {
>           await windowManager.setTitle("Child Window")
>           await windowManager.show()
>       })
>    }
> }
> ```

When `DesktopMultiWindow.createWindow(...)` is called, `void main(args)` is run, in my case is not work. only call 'DesktopMultiWindow.createWindow(...).then(window) { window.show() } ' that `void main(args)` can run

**Comment by isaacy13** [comment]
@zjw78123sc Thanks, ran into this issue recently trying to do it for macOS

Now, although the child window is created, it doesn't seem to be registering the plugins properly

Even though it's supposedly registering plugins for new window's view controller, it's still giving errors for multiple libraries: `implementation is not found`

Have you run into this issue?
Flutter v3.16.8 - Dart 3.2.5 - macOS Sequoia 15.3

```
import Cocoa
import FlutterMacOS
import desktop_multi_window

class MainFlutterWindow: NSWindow {
  override func awakeFromNib() {
    let flutterViewController = FlutterViewController.init()
    let windowFrame = self.frame
    self.contentViewController = flutterViewController
    self.setFrame(windowFrame, display: true)

    RegisterGeneratedPlugins(registry: flutterViewController)
    
    FlutterMultiWindowPlugin.setOnWindowCreatedCallback { flutterVC in
      RegisterGeneratedPlugins(registry: flutterVC)
    }

    super.awakeFromNib()
  }
}
```
